# Yellow Taxi Trip Data 2017

In [17]:
import pandas as pd              
import numpy as np 

In [2]:
df = pd.read_csv('2017_Yellow_Taxi_Trip_Data.csv')
print("Done")


Done


## Inspecting the data

In [3]:
df.head(10)


Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
0    24870114         2   03/25/2017 8:55:43 AM   03/25/2017 9:09:47 AM   
1    35634249         1   04/11/2017 2:53:28 PM   04/11/2017 3:19:58 PM   
2   106203690         1   12/15/2017 7:26:56 AM   12/15/2017 7:34:08 AM   
3    38942136         2   05/07/2017 1:17:59 PM   05/07/2017 1:48:14 PM   
4    30841670         2  04/15/2017 11:32:20 PM  04/15/2017 11:49:03 PM   
5    23345809         2   03/25/2017 8:34:11 PM   03/25/2017 8:42:11 PM   
6    37660487         2   05/03/2017 7:04:09 PM   05/03/2017 8:03:47 PM   
7    69059411         2   08/15/2017 5:41:06 PM   08/15/2017 6:03:05 PM   
8     8433159         2   02/04/2017 4:17:07 PM   02/04/2017 4:29:14 PM   
9    95294817         1   11/10/2017 3:20:29 PM   11/10/2017 3:40:55 PM   

   passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
0                6           3.34           1                  N   
1                1           1.80           1                  N   
2                1           1.00           1                  N   
3                1           3.70           1                  N   
4                1           4.37           1                  N   
5                6           2.30           1                  N   
6                1          12.83           1                  N   
7                1           2.98           1                  N   
8                1           1.20           1                  N   
9                1           1.60           1                  N   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
0           100           231             1         13.0    0.0      0.5   
1           186            43             1         16.0    0.0      0.5   
2           262           236             1          6.5    0.0      0.5   
3           188            97             1         20.5    0.0      0.5   
4             4           112             2         16.5    0.5      0.5   
5           161           236             1          9.0    0.5      0.5   
6            79           241             1         47.5    1.0      0.5   
7           237           114             1         16.0    1.0      0.5   
8           234           249             2          9.0    0.0      0.5   
9           239           237             1         13.0    0.0      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  
0        2.76           0.0                    0.3         16.56  
1        4.00           0.0                    0.3         20.80  
2        1.45           0.0                    0.3          8.75  
3        6.39           0.0                    0.3         27.69  
4        0.00           0.0                    0.3         17.80  
5        2.06           0.0                    0.3         12.36  
6        9.86           0.0                    0.3         59.16  
7        1.78           0.0                    0.3         19.58  
8        0.00           0.0                    0.3          9.80  
9        2.75           0.0                    0.3         16.55

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22699 entries, 0 to 22698
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             22699 non-null  int64  
 1   VendorID               22699 non-null  int64  
 2   tpep_pickup_datetime   22699 non-null  object 
 3   tpep_dropoff_datetime  22699 non-null  object 
 4   passenger_count        22699 non-null  int64  
 5   trip_distance          22699 non-null  float64
 6   RatecodeID             22699 non-null  int64  
 7   store_and_fwd_flag     22699 non-null  object 
 8   PULocationID           22699 non-null  int64  
 9   DOLocationID           22699 non-null  int64  
 10  payment_type           22699 non-null  int64  
 11  fare_amount            22699 non-null  float64
 12  extra                  22699 non-null  float64
 13  mta_tax                22699 non-null  float64
 14  tip_amount             22699 non-null  float64
 15  to

In [5]:
df.describe()

Unnamed: 0      VendorID  passenger_count  trip_distance  \
count  2.269900e+04  22699.000000     22699.000000   22699.000000   
mean   5.675849e+07      1.556236         1.642319       2.913313   
std    3.274493e+07      0.496838         1.285231       3.653171   
min    1.212700e+04      1.000000         0.000000       0.000000   
25%    2.852056e+07      1.000000         1.000000       0.990000   
50%    5.673150e+07      2.000000         1.000000       1.610000   
75%    8.537452e+07      2.000000         2.000000       3.060000   
max    1.134863e+08      2.000000         6.000000      33.960000   

         RatecodeID  PULocationID  DOLocationID  payment_type   fare_amount  \
count  22699.000000  22699.000000  22699.000000  22699.000000  22699.000000   
mean       1.043394    162.412353    161.527997      1.336887     13.026629   
std        0.708391     66.633373     70.139691      0.496211     13.243791   
min        1.000000      1.000000      1.000000      1.000000   -120.000000   
25%        1.000000    114.000000    112.000000      1.000000      6.500000   
50%        1.000000    162.000000    162.000000      1.000000      9.500000   
75%        1.000000    233.000000    233.000000      2.000000     14.500000   
max       99.000000    265.000000    265.000000      4.000000    999.990000   

              extra       mta_tax    tip_amount  tolls_amount  \
count  22699.000000  22699.000000  22699.000000  22699.000000   
mean       0.333275      0.497445      1.835781      0.312542   
std        0.463097      0.039465      2.800626      1.399212   
min       -1.000000     -0.500000      0.000000      0.000000   
25%        0.000000      0.500000      0.000000      0.000000   
50%        0.000000      0.500000      1.350000      0.000000   
75%        0.500000      0.500000      2.450000      0.000000   
max        4.500000      0.500000    200.000000     19.100000   

       improvement_surcharge  total_amount  
count           22699.000000  22699.000000  
mean                0.299551     16.310502  
std                 0.015673     16.097295  
min                -0.300000   -120.300000  
25%                 0.300000      8.750000  
50%                 0.300000     11.800000  
75%                 0.300000     17.800000  
max                 0.300000   1200.290000

## Investigating the variables

# Sort the data by trip distance from maximum to minimum value


In [6]:
df_sort = df.sort_values(by=['trip_distance'],ascending=False)
df_sort.head(10)

Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
9280     51810714         2  06/18/2017 11:33:25 PM  06/19/2017 12:12:38 AM   
13861    40523668         2   05/19/2017 8:20:21 AM   05/19/2017 9:20:30 AM   
6064     49894023         2  06/13/2017 12:30:22 PM   06/13/2017 1:37:51 PM   
10291    76319330         2  09/11/2017 11:41:04 AM  09/11/2017 12:18:58 PM   
29       94052446         2   11/06/2017 8:30:50 PM  11/07/2017 12:00:00 AM   
18130    90375786         1   10/26/2017 2:45:01 PM   10/26/2017 4:12:49 PM   
5792     68023798         2   08/11/2017 2:14:01 PM   08/11/2017 3:17:31 PM   
15350    77309977         2   09/14/2017 1:44:44 PM   09/14/2017 2:34:29 PM   
10302    43431843         1   05/15/2017 8:11:34 AM   05/15/2017 9:03:16 AM   
2592     51094874         2   06/16/2017 6:51:20 PM   06/16/2017 7:41:42 PM   

       passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
9280                 2          33.96           5                  N   
13861                1          33.92           5                  N   
6064                 1          32.72           3                  N   
10291                1          31.95           4                  N   
29                   1          30.83           1                  N   
18130                1          30.50           1                  N   
5792                 1          30.33           2                  N   
15350                1          28.23           2                  N   
10302                1          28.20           2                  N   
2592                 1          27.97           2                  N   

       PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
9280            132           265             2       150.00    0.0      0.0   
13861           229           265             1       200.01    0.0      0.5   
6064            138             1             1       107.00    0.0      0.0   
10291           138           265             2       131.00    0.0      0.5   
29              132            23             1        80.00    0.5      0.5   
18130           132           220             1        90.50    0.0      0.5   
5792            132           158             1        52.00    0.0      0.5   
15350            13           132             1        52.00    0.0      0.5   
10302            90           132             1        52.00    0.0      0.5   
2592            261           132             2        52.00    4.5      0.5   

       tip_amount  tolls_amount  improvement_surcharge  total_amount  
9280         0.00          0.00                    0.3        150.30  
13861       51.64          5.76                    0.3        258.21  
6064        55.50         16.26                    0.3        179.06  
10291        0.00          0.00                    0.3        131.80  
29          18.56         11.52                    0.3        111.38  
18130       19.85          8.16                    0.3        119.31  
5792        14.64          5.76                    0.3         73.20  
15350        4.40          5.76                    0.3         62.96  
10302       11.71          5.76                    0.3         70.27  
2592         0.00          5.76                    0.3         63.06

When we look at the sorted results for the trip distances, the longest ride in the dataset is about 33.96 miles, and many of the other top values are also between 30 and 33 miles. Normally, most taxi rides in New York City are much shorter, usually just 1 to 3 miles within the city. However, longer trips around 30 miles are realistic because passengers may be traveling to airports like JFK or LaGuardia, or even going to nearby suburbs outside Manhattan. Since there are no extreme values such as trips over 100 miles, which would be suspicious for a city taxi, these distances appear to be reasonable and make sense in the context of New York City taxi rides.

**Question 1: Sort your first variable (`trip_distance`) from maximum to minimum value, do the values seem normal?**

Yes, the values seem normal. The maximum trip distance is about 34 miles, which makes sense for New York City since long trips (like to airports or nearby towns) can reach this range. Most trips are short, but these longer trips are reasonable and not unusual.

# Sorting the data by total amount and print the top 20 values


In [7]:
# Sorting the data by total amount and print the top 20 values

total_amount_sorted = df.sort_values(
    ['total_amount'], ascending=False)['total_amount']

total_amount_sorted.head(20)


8476     1200.29
20312     450.30
13861     258.21
12511     233.74
15474     211.80
6064      179.06
16379     157.06
3582      152.30
11269     151.82
9280      150.30
1928      137.80
10291     131.80
6708      126.00
11608     123.30
908       121.56
7281      120.96
18130     119.31
13621     115.94
13359     111.95
29        111.38
Name: total_amount, dtype: float64

**Question 2: Sort by your second variable (`total_amount`), are any values unusual?**

When we sort the dataset by the **`total_amount`** column, we notice that some values appear unusual. For example, the highest total amount is **\$1200.29**, and there are other rides costing **\$450**, **\$258**, and **\$233**. These amounts are far above the normal range for most New York City taxi rides, which usually cost between **\$10** and **\$50** depending on distance, tolls, and tips. While it is possible that some very expensive trips occurred—such as long-distance rides or cases with high tolls—the extremely high fare of over **\$1200** looks unrealistic and likely represents an error or outlier in the data. Therefore, although most values are reasonable, the top few unusually large amounts should be flagged for further investigation or cleaning.




# Sort the data by total amount and print the bottom 20 values

In [8]:
# Sort the data by total amount and print the bottom 20 values
total_amount_sorted.tail(20)

14283      0.31
19067      0.30
10506      0.00
5722       0.00
4402       0.00
22566      0.00
1646      -3.30
18565     -3.80
314       -3.80
5758      -3.80
5448      -4.30
4423      -4.30
10281     -4.30
8204      -4.80
20317     -4.80
11204     -5.30
14714     -5.30
17602     -5.80
20698     -5.80
12944   -120.30
Name: total_amount, dtype: float64

**Question 3: Are the resulting rows similar for both sorts? Why or why not?**

When we compare the rows sorted by trip_distance and total_amount, we see that they are not the same. The longest trips are not always the most expensive ones. This happens because the total_amount also depends on other factors like tips, tolls, surcharges, and sometimes errors in the data. For example, a short trip with a very large tip or tolls may appear more expensive than a much longer trip. This means distance and fare are related, but not perfectly proportional.

# How many of each payment type are represented in the data?

In [9]:
# How many of each payment type are represented in the data?
df['payment_type'].value_counts()


payment_type
1    15265
2     7267
3      121
4       46
Name: count, dtype: int64

This output shows the counts of different `payment_type` values in the dataset:

- **1 → 15,265 rides** (this usually means **Credit Card**)  
- **2 → 7,267 rides** (this usually means **Cash**)  
- **3 → 121 rides** (often represents **No charge**)  
- **4 → 46 rides** (often represents **Dispute** or **Other**)

So, most passengers paid by **credit card**, followed by **cash**, while very few trips had “no charge” or “disputed” payments.


The data shows how passengers paid for their taxi rides. Most people used **credit cards** (about 15,265 rides), while the second most common method was **cash** (around 7,267 rides). A very small number of rides were marked as **no charge** (121 rides), which could mean free trips or errors in the system. Finally, only **46 rides** were recorded as **disputed payments**, which means there was some problem or disagreement with the fare.  

Overall, credit card and cash are the main ways people paid, while the other methods are very rare.


# What is the average tip for trips paid for with credit card?

In [10]:
# What is the average tip for trips paid for with credit card?
avg_cc_tip = df[df['payment_type']==1]['tip_amount'].mean()
print('Avg. cc tip:', avg_cc_tip)

Avg. cc tip: 2.7298001965280054


# What is the average tip for trips paid for with cash?

In [11]:
# What is the average tip for trips paid for with cash?
avg_cash_tip = df[df['payment_type']==2]['tip_amount'].mean()
print('Avg. cash tip:', avg_cash_tip)

Avg. cash tip: 0.0


The average tip for credit card rides is about 2.73, while the average tip for cash rides is 0.0. This happens because credit card tips are recorded automatically in the system, but cash tips are usually given directly to the driver and not entered into the dataset. As a result, credit card rides show recorded tips, while cash rides mostly appear as zero.

# How many times is each vendor ID represented in the data?

In [12]:
# How many times is each vendor ID represented in the data?
df['VendorID'].value_counts()

VendorID
2    12626
1    10073
Name: count, dtype: int64

# What is the mean total amount for each vendor?

In [13]:
# What is the mean total amount for each vendor?
df.groupby(['VendorID']).mean(numeric_only=True)[['total_amount']]


total_amount
VendorID              
1            16.298119
2            16.320382

There are 12,626 rides recorded by Vendor 2 and 10,073 rides recorded by Vendor 1. This means Vendor 2 has slightly more trips in the dataset. However, when comparing the average total fare, both vendors are very similar: Vendor 1 has an average fare of about 16.30, and Vendor 2 also has an average of about 16.32. This shows that while the number of trips differs, the average fare charged by both vendors is nearly the same.

# Filter the data for credit card payments only

In [24]:
# Filter the data for credit card payments only
credit_card = df[df['payment_type']==1]


# Filter the credit-card-only data for passenger count only

In [15]:
# Filter the credit-card-only data for passenger count only
credit_card['passenger_count'].value_counts()

passenger_count
1    10977
2     2168
5      775
3      600
6      451
4      267
0       27
Name: count, dtype: int64

Most credit card trips are taken by a single passenger (10,977 rides). The second most common group is two passengers (2,168 rides). Larger groups, such as 3, 4, 5, or 6 passengers, appear less often. There are also 27 cases with 0 passengers, which seems unusual and likely represents errors in the data. Overall, the majority of credit card payments come from rides with just one or two people.

# Calculate the average tip amount for each passenger count (credit card payments only)

In [16]:
# Calculate the average tip amount for each passenger count (credit card payments only)
credit_card.groupby(['passenger_count']).mean(numeric_only=True)[['tip_amount']]

tip_amount
passenger_count            
0                  2.610370
1                  2.714681
2                  2.829949
3                  2.726800
4                  2.607753
5                  2.762645
6                  2.643326

- The average tip amount for credit card rides is quite consistent across all passenger counts, staying around 2.6 to 2.8.

- The highest average tips are for 2 passengers (2.83) and 5 passengers (2.76).

- The lowest average tips are for 4 passengers (2.61) and rides with 0 passengers (2.61), which may reflect data errors.

- Overall, the number of passengers does not have a strong effect on the average tip amount, since the values are very similar across all groups.

# According to Luana's requests, which two variables are most helpful for building a predictive model for the client: NYC TLC?

After analyzing the dataset, the two most useful variables for building a predictive model for NYC taxi ride fares are **`trip_distance`** and **`total_amount`**. The reason is that `trip_distance` explains how far the passenger traveled, and `total_amount` reflects the cost of the ride including fare, taxes, and tips. These two variables give the clearest picture of a taxi trip and are directly connected to how fares are calculated. Other factors like passenger count or payment type provide extra context, but they are less important for predicting the ride fare compared to distance and total amount.
